rust/rope/src/spans.rs

```diff
@@ -189,6 +189,26 @@ impl<T: Clone + Default> Spans<T> {
             ix: 0,
         }
     }
+
+    pub fn clear_left(&mut self, pos: usize) {
+        let iv = Interval::new_closed_open(0, pos);
+        let mut b = TreeBuilder::new();
+        b.push(SpansBuilder::new(iv.end()).build());
+        self.push_subseq(&mut b, self.interval().suffix(iv));
+        *self = b.build();
+    }
+
+    pub fn shift_left(&mut self, len: usize) {
+        let iv = Interval::new_closed_open(len, self.interval().end());
+        *self = self.subseq(iv);
+    }
+
+    pub fn shift_right(&mut self, len: usize) {
+        let mut b = TreeBuilder::new();
+        b.push(SpansBuilder::new(len).build());
+        self.push_subseq(&mut b, self.interval());
+        *self = b.build();
+    }
```

Comment:
To be honest, I don't completely understand the tree API, yet. While these methods are doing what I am intending to do, I have the feeling that their implementation could somehow be improved a lot?